demos/camassa-holm/camassaholm.py.rst

```diff
@@ -0,0 +1,203 @@
+Camassa-Holm equation
+=====================
+
+The Camassa Holm equation [CH1993]_ is an integrable 1+1 PDE which may be
+written in the form
+
+.. math::
+   m_t + mu_x + (mu)_x = 0, \quad u - \alpha^2u_{xx} = m,
+
+solved in the interval :math:`[a,b]` either with periodic boundary
+conditions or with boundary conditions `u(a)=u(b)=0`; :math:`alpha>0`
+is a constant that sets a lengthscale for the solution. The solution
+is entirely composed of peaked solitons corresponding to Dirac delta
+functions in :math:`m`. Further, the solution has a conserved energy,
+given by
+
+.. math::
+   \int_a^b \frac{1}{2} u^2 + \frac{\alpha^2}{2} u_x^2 \mathrm{d}x.
+
+In this example we will concentrate on the periodic boundary
+conditions case.
+
+A weak form of these equations is given by
+
+.. math::
+   \int pm_t + pmu_x - p_xmu \mathrm{d}x=0, \, \forall p,
+   
+   \int qu + \alpha^2q_xu_x - qm \mathrm{d}x=0, \forall q.
+
+Energy conservation then follows from substituting the second equation
+into the first, and then setting :math:`p=u',
+
+.. math::
+   \dot{E} &= \frac{\mathrm{d}}{\mathrm{d}t}\int_a^b \frac{1}{2}u^2 + \frac{\alpha^2}{2}u_x^2 \diff x \\
+   &= \int_a^b uu_t + \alpha^2 u_xu_{xt} \diff x, \\
+   &= \int_a^b um_t \diff x, \\
+   &= \int_a^b -umu_x + u_xmu \diff x = 0.
+
+If we choose the same continuous finite element spaces for $m$ and $u$
+then this proof immediately extends to the spatial discretisation, as
+noted by [Ma2010]_. Further, it is a property of the implicit midpoint
+rule time discretisation that any quadratic conserved quantities of an
+ODE are also conserved by the time discretisation (see [Is2009]_, for
+example). Hence, the fully discrete scheme,
+
+.. math::
+   \int p(m^{n+1}-m^n} + \Delta t(pm^{n+1/2}u^{n+1/2}_x - p_xm^{n+1/2}u^{n+1/2}) \mathrm{d}x=0, \, \forall p\in V,
+   
+   \int qu + \alpha^2q_xu_x - qm \mathrm{d}x=0, \forall q \in V,
+
+where :math:`u^{n+1/2}=(u^{n+1}+u^n)/2`,
+:math:`m^{n+1/2}=(m^{n+1}+m^n)/2`, conserves the energy exactly. This
+is a useful property since the energy is the square of the :math:`H^1`
+norm, which guarantees regularity of the numerical solution.
+
+To implement this, the following code is currently required to ensure
+reevaluation of the energy for output.::
+
+  from firedrake import *
+  disable_performance_optimisations()
```

Comment:
I dunno. I assumed that the reason disable_performance_optimisations() was
introduced was precisely to avoid the user having to know how to switch
individual ones off. It's a bug.

On 13 December 2015 at 10:58, Andrew T. T. McRae <notifications@github.com>
wrote:

> In demos/camassa-holm/camassaholm.py.rst
> <https://github.com/firedrakeproject/firedrake/pull/662#discussion_r47443033>
> :
>
> > +
> > +.. math::
> > +   \int p(m^{n+1}-m^n} + \Delta t(pm^{n+1/2}u^{n+1/2}_x - p_xm^{n+1/2}u^{n+1/2}) \mathrm{d}x=0, \, \forall p\in V,
> > +
> > +   \int qu + \alpha^2q_xu_x - qm \mathrm{d}x=0, \forall q \in V,
> > +
> > +where :math:`u^{n+1/2}=(u^{n+1}+u^n)/2`,
> > +:math:`m^{n+1/2}=(m^{n+1}+m^n)/2`, conserves the energy exactly. This
> > +is a useful property since the energy is the square of the :math:`H^1`
> > +norm, which guarantees regularity of the numerical solution.
> > +
> > +To implement this, the following code is currently required to ensure
> > +reevaluation of the energy for output.::
> > +
> > +  from firedrake import *
> > +  disable_performance_optimisations()
>
> Is that really necessary, or can you just disable assembly cache?
> parameters["assembly_cache"]["enabled"] = False
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/firedrakeproject/firedrake/pull/662/files#r47443033>.
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916